### PR TITLE
Attemted fix of #9890 mouse lock when dragging and losing focus

### PR DESCRIPTION
**Pull request by passeride**
Junior here so probably needs review. 

But bug was fixed in scenario specified by #9890 

edit: first pullreq, will read code guide and resubmit

### CODE REVIEW

File: scene/gui/tree.cpp [hunk 1]
```diff
@@ -28,6 +28,7 @@
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
 /*************************************************************************/
 #include "tree.h"
+#include "../../core/os/main_loop.h"
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: `core` is already in the include path, you can use `os/main_loop.h` directly. Also best put it next to the other Godot-related includes. Not sure why `<limits.h>` is there, but it should actually come last:
```
#include "tree.h"

#include "os/input.h"
#include "os/keyboard.h"
#include "os/main_loop.h"
#include "os/os.h"
#include "print_string.h"
#include "project_settings.h"
#include "scene/main/viewport.h"

#include <limits.h>
```

File: scene/gui/tree.cpp [hunk 2]
```diff
@@ -2448,7 +2458,14 @@ void Tree::_gui_input(Ref<InputEvent> p_event) {
 	}
 
 	Ref<InputEventMouseButton> b = p_event;
-
+	
+	if(is_dragExited()){
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: Wrong indentation and style, see http://docs.godotengine.org/en/latest/community/contributing/code_style_guidelines.html

File: scene/gui/tree.cpp [hunk 3]
```diff
@@ -134,6 +135,14 @@ TreeItem::TreeCellMode TreeItem::get_cell_mode(int p_column) const {
 	return cells[p_column].mode;
 }
 
+void Tree::set_dragExited(bool b){
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: As you can see, all methods of the Godot API use snake_case, so you should do the same.

File: scene/gui/tree.h [hunk 4]
```diff
@@ -178,6 +178,8 @@ class TreeItem : public Object {
 	}
 
 public:
+  
+
```
Comment (akien-mga) [inline scene/gui/tree.h]: Style issue.

File: scene/gui/tree.h [hunk 5]
```diff
@@ -524,6 +527,11 @@ class Tree : public Control {
 	}
 
 public:
+  
+  	/* Test for focus loss */
+	void set_dragExited(bool b);
```
Comment (akien-mga) [inline scene/gui/tree.h]: You should use meaningful parameter names, one letter `b` doesn't say anything about its purpose.
As you can see in other methods of this class and all over the Godot API, parameters should be prefixed with `p_`.

But actually, I don't see the point of those public methods. `drag_exited` only needs to be a private bool, you can change it directly from within the class. The setter and getter don't add any value.

**Comment by akien-mga** [comment]
Welcome aboard! This is a good start, but there are various things to fix to better match Godot's style. The main rule is that you should use the same style as the rest of the class you're editing, so read its existing code a bit and it should be relatively obvious. I've made some inline comments to point out some issues.

Once ready you'll have to rebase your branch to squash all commits together. There are instructions to do so in the [PR workflow](http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html) documentation.

File: scene/gui/tree.cpp [hunk 6]
```diff
@@ -2447,6 +2449,14 @@ void Tree::_gui_input(Ref<InputEvent> p_event) {
 		}
 	}
 
+  if (drag_exited) {
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: Indentation issue (same for closing bracket).

File: scene/gui/tree.cpp [hunk 7]
```diff
@@ -2933,6 +2943,14 @@ void Tree::_notification(int p_what) {
 	if (p_what == NOTIFICATION_THEME_CHANGED) {
 		update_cache();
 	}
+
+	if (p_what == MainLoop::NOTIFICATION_WM_FOCUS_IN) {
+		drag_exited = false;
+	}
+
+	if (p_what == MainLoop::NOTIFICATION_WM_FOCUS_OUT) {
+		drag_exited = true;;
```
Comment (akien-mga) [inline scene/gui/tree.cpp]: There's an extraneous semicolon here.

**Comment by akien-mga** [comment]
You properly squashed the commits, though you changed the resulting commit log to say "Cleaning up code as pr review", but:
- That's not really useful information for the commit history
- That doesn't say anything about what the commit actually does :D

File: scene/gui/tree.cpp (hunk 6 above)
Comment (reduz) [inline scene/gui/tree.cpp]: is there a reason why this cant be like

if ( condition_for_dragging && ( p_what== NOTIFICATION_WM_FOCUS_IN ||  p_what== NOTIFICATION_WM_FOCUS_OUT)) {
     disable dragging?
}

Comment (passeride) [reply scene/gui/tree.cpp]: p_what is a parameter for _notification function, and i think to cache the event in a variable is needed.  But i might be wrong..

**Comment by passeride** [comment]
Checks seems to be affected by a bug introduces in 65fb961 regarding thekla_atlas making problems during compilation

File: scene/gui/tree.cpp [hunk 8]
```diff
@@ -2449,13 +2449,13 @@ void Tree::_gui_input(Ref<InputEvent> p_event) {
 		}
 	}
 
-  if (drag_exited) {
+	if (drag_exited) {
```
Comment (reduz) [inline scene/gui/tree.cpp]: What i dont understand is why this code cant be in NOTIFICATION_WM_FOCUS out, did you try it and it didnt work?

**Comment by akien-mga** [comment]
Thanks a lot for your contribution!
We are now entering a strict release freeze for Godot 3.0 and will only consider major bug fixes. We won't merge new features and enhancements anymore until 3.0 is released.

Moving this PR to the 3.1 milestone, to be reviewed once the release freeze is lifted. It could eventually be cherry-picked for a future 3.0.1 maintenance release if it doesn't change the user-facing APIs and doesn't compromise the engine's stability.

**Comment by akien-mga** [comment]
This was superseded by #19351, but thanks for the contribution!

**Pull request closed without merging**